### PR TITLE
Cloud browser support

### PR DESCRIPTION
**Pull request by gbartoloni**
Added support for the automation to run on cloud browser 'browserstack'
Please provide feedbacks on the general rspec structure and ruby implementation. In this repo both test frameworks 'rspec' and 'cucumber' are sharing the same classes and page objects to optimize the code and yet leave the team to use either one depending on their needs.

### CODE REVIEW

File: test_commons/app/pages/login.rb [hunk 1]
```diff
@@ -2,14 +2,22 @@
 module App
     module Pages
         class Login <SitePrism::Page
-            # @login_page.visit("http://user@pass.integration.ampion.net")
-            # set_url "https://portal.integration.ampion.net/users/sign_in#/"
+            test_env = ENV['TEST_ENV'] || 'local'
+            url = (test_env == 'local' ? "http://localhost:3000/users/sign_in#/" : 
+            "https://portal.integration.ampion.net/users/sign_in#/")
```
Comment (seanhussey) [inline test_commons/app/pages/login.rb]: Indent this line just a bit (one tab) so it looks like a continuation of the line above it.

File: spec/rspec_helpers.rb [hunk 2]
```diff
@@ -9,17 +10,64 @@
...
-Capybara.reset_sessions!
-Capybara.current_session.driver.browser.manage.window.resize_to(1000, 800)
-        
-$page_set||=App::Pages::PageCollection.new
-# $section_set||=App::Sections::SectionCollection.new
 
+if browser_location == 'local'
+  
+  Capybara.register_driver :selenium do |app|
+      client = Selenium::WebDriver::Remote::Http::Default.new
```
Comment (seanhussey) [inline spec/rspec_helpers.rb]: These 3 lines are indented too far.

File: spec/rspec_helpers.rb [hunk 3]
```diff
@@ -9,17 +10,64 @@
...
+      client.timeout =  30
+      Capybara::Selenium::Driver.new(app, :browser => :chrome, :http_client => client)
+  end
+  Capybara.default_driver = :selenium
+  Capybara.current_session.driver.browser.manage.window.resize_to(1000, 800)
+
+elsif browser_location == 'browserstack'
+  # monkey patch to avoid reset sessions
```
Comment (seanhussey) [inline spec/rspec_helpers.rb]: What does this comment refer to?

File: spec/rspec_helpers.rb [hunk 4]
```diff
@@ -9,17 +10,64 @@
...
+
+
```
Comment (seanhussey) [inline spec/rspec_helpers.rb]: Remove extra blank line.

File: config/browserstack.config.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+browser_caps:
+  -
+    "browser": "chrome"
+    # "browser_version": "11.0"
+    # "os": "Windows"
+    # "os_version": "10"
+    # "resolution": "1024x768"
+browser_url: "https://portal.integration.ampion.net/users/sign_in#/"
+superadmin_username: "sue@example.com"
+superadmin_password: "Mypassword123"
```
Comment (seanhussey) [inline config/browserstack.config.yml]: Add a newline.

File: features/superadmin_create_contract.feature [hunk 6]
```diff
@@ -0,0 +1,6 @@
+Feature: Superadmin Creates a new Contract
+
+    Scenario: As a Superadmin user I can add a new Contract that requires a credit check
+    Scenario: As a Client Service user I can add a new Contract that requires a credit check
+    Scenario: As a Superadmin I can edit an existing contract and toggle the credit check checkbox
+    Scenario: As a Superadmin I cannot edit a contract credit check if there is at least one account using it
```
Comment (seanhussey) [inline features/superadmin_create_contract.feature]: Add a newline.

File: features/user_access/clientservice_access_rights.feature [hunk 7]
```diff
@@ -0,0 +1,10 @@
+Feature: Client Service User access rights
+    Scenario: Client Service User Adds an operator
+    Scenario: Client Service User creates a Customer Representative
+    Scenario: Client service User cannot create users or utility rates
+    Scenario: Client Service User Impersonate an operator
+    Scenario: Client Service User create Operator Admin
+    Scenario: Client Service User creates Operator User
+    Scenario: Client Service User creates Maintenance User
+    Scenario: Client Service User creates Customer User
+    Scenario: Client Service User creates Guest User
```
Comment (seanhussey) [inline features/user_access/clientservice_access_rights.feature]: Add a newline.

File: features/user_access/customer_access_rights.feature [hunk 8]
```diff
@@ -0,0 +1,2 @@
+Feature: Customer User access rights
+Scenario: Customer User creates Guest User
```
Comment (seanhussey) [inline features/user_access/customer_access_rights.feature]: Add a newline.

File: features/user_access/custservice_access_rights.feature [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Feature: Customer Service User access Rights
+
+    Scenario: Customer Service User creates Customer User
+    Scenario: Customer Service User creates Guest User
+    Scenario: Customer Service usesr Impersonates Operator
```
Comment (seanhussey) [inline features/user_access/custservice_access_rights.feature]: Add a newline.

File: features/user_access/operator_access_rights.feature [hunk 10]
```diff
@@ -0,0 +1,9 @@
+Feature: Operator User access Rights
+
+    Scenario: Operator Admin creates SCenario: Operator Admin
+    Scenario: Operator Admin creates Operator User
+    Scenario: Operator Admin creates Maintenance User
+    Scenario: Operator Admin creates Customer User
+    Scenario: Operator Admin creates Guest User
+    Scenario: Operator User creates Customer User
+    Scenario: Operator User creates Guest User
```
Comment (seanhussey) [inline features/user_access/operator_access_rights.feature]: Add a newline.

File: features/user_access/superadmin_access_rights.feature [hunk 11]
```diff
@@ -0,0 +1,9 @@
+Feature: Superadmin access and rights
+    Scenario: Superadmin creates Operator Admin
+    Scenario: Superadmin creates Operator User
+    Scenario: Superadmin creates Maintenance User
+    Scenario: Superadmin creates Client Service user
+    Scenario: Superadmin Creates Scenario: Superadmin
+    Scenario: Superadmin Impersonate Customer
+    Scenario: Superadmin creates Customer User (via operator)
+    Scenario: Superadmin creates Guest User
```
Comment (seanhussey) [inline features/user_access/superadmin_access_rights.feature]: Add a newline.

File: spec/rspec_helpers.rb [hunk 12]
```diff
@@ -9,17 +10,78 @@
...
+    )                                                                              
+    driver = Capybara::Selenium::Driver.new(                                       
+      app,                                                                         
+      browser: :chrome,                                                            
+      desired_capabilities: caps                                                   
+    )                                                                              
+  end                                                                              
+  Capybara.default_driver = :headless_chromium
+else
+  raise("Invalid browser_location env var Specified. [local]||browserstack")
```
Comment (vrogers) [inline spec/rspec_helpers.rb]: probably want to add || chromium here right?

**Comment by vrogers** [comment]
I'll have to get familiar with this technology over time to provide more informed feedback, but it looks good overall!

File: test_commons/app/pages/login.rb (hunk 1 above)
Comment (gbartoloni) [reply test_commons/app/pages/login.rb]: Will do.

File: spec/rspec_helpers.rb (hunk 3 above)
Comment (gbartoloni) [reply spec/rspec_helpers.rb]: outdated, thank you for point it out.

File: spec/rspec_helpers.rb (hunk 12 above)
Comment (gbartoloni) [reply spec/rspec_helpers.rb]: Yes good catch.

File: config/browserstack.config.yml (hunk 5 above)
Comment (gbartoloni) [reply config/browserstack.config.yml]: Thanks , adding new line here and to the other files

**Pull request merged**